torch/_functorch/aot_autograd.py

```diff
@@ -2801,6 +2812,19 @@ def functional_call(*args, **kwargs):
     assert static_argnums is None
     if bw_compiler is None:
         bw_compiler = fw_compiler
+
+    full_args = []
+    full_args.extend(params_flat)
+    full_args.extend(args)
+
+    dynamic_shapes = False
+    for x in full_args:
+        if isinstance(x, FakeTensor):
+            dynamic_shapes = x.fake_mode.shape_env is not None
```

Comment:
Should we assert that this is all or nothing? (that either all fake tensors share the same shape env, or none of them have one?)

Reply:
Why is this change needed?

Reply:
The proximal reason for this change is, the way I decided to configure if you do AOTAutograd with dynamic shapes or not was add an entry to AOTConfig. This means that anywhere I create an AOTConfig, I need to fill in this field, if I want dynamic shapes that are not controlled by the functorch config.

In particular, when aot_module_simplified is invoked, the way we tell if we were running with dynamo dynamic shapes (because we don't have access to torch._dynamo.config, due to layering problems) is by inspecting the input arguments and checking if any of them are fake. This logic is reduxed with `create_aot_dispatcher_function` which also does a similar sniffing (and a further refactor would be to extract the ShapeEnv and then eliminate the equivalent logic in `create_aot_dispatcher_function`).

If I didn't do this change, then when we compile backwards (where we pass in all real tensors), we would not compile with dynamic shapes. I do think this part is still a little creaky though.